src/librustc_mir/shim.rs

```diff
@@ -66,9 +66,12 @@ fn make_shim<'tcx>(tcx: TyCtxt<'tcx>, instance: ty::InstanceDef<'tcx>) -> &'tcx
                 Some(arg_tys)
             )
         }
-        ty::InstanceDef::Virtual(def_id, _) => {
-            // We are generating a call back to our def-id, which the
-            // codegen backend knows to turn to an actual virtual call.
+        // We are generating a call back to our def-id, which the
+        // codegen backend knows to turn to an actual virtual call.
+        ty::InstanceDef::Virtual(def_id, _) |
+        // ...or we are generating a direct call to our #[track_caller] function which
+        // requires an implicit caller location that the virtual call won't pass
```

Comment:
s/virtual/indirect (to distinguish from `dyn Trait` method calls).

I'd still like this to be given as an example of a more general pattern (i.e. direct call needs to be codegen'd differently than an indirect call), because the shim is called `ReifyShim` not `TrackCallerShim`.